### ISSUE TITLE
Type of `typing.Any` should be `typing._AnyMeta`, not `typing.Any`

### ISSUE THREAD

**Issue by IAmREGE** [microsoft/pyright#7787]
**Describe the bug**
Run `pyright Untitled-1.py` in a shell.

**Expected Behaviour**
Output `Untitled-1.py:3:13 - information: Type of "Any" is "_AnyMeta"` like Runtime Python. (Mypy returns `builtins.object`.)

**Actual Behaviour**
Output `Untitled-1.py:3:13 - information: Type of "Any" is "Any"`.

**Code or Screenshots**

**`Untitled-1.py`**
```python
from typing import Any, reveal_type

reveal_type(Any)
```

**VS Code extension or command-line**
Pyright 1.1.360, installed from pacman.

**Comment by erictraut** [comment]
Pyright's behavior here is correct and consistent with the Python typing spec. The type of `Any` is `Any`. It's a special form that has a specific meaning in the type system. You're confusing the type with its internal implementation at runtime.

**Comment by hexane360** [comment]
I'm running into this as well. Currently, I don't think there's any way to type an expression that contains the literal value `Any`.

I'm sorry if this is a basic question, but can you specify where it's specified that the type of `Any` is `Any`? What is the logic behind this decision?

The stdlib `typing` documentation just says:

>A special kind of type is [Any](https://docs.python.org/3/library/typing.html#typing.Any). A static type checker will treat every type as being compatible with [Any](https://docs.python.org/3/library/typing.html#typing.Any) and [Any](https://docs.python.org/3/library/typing.html#typing.Any) as being compatible with every type.

And PEP 484 just says:

>A special kind of type is Any. Every type is consistent with Any. It can be considered a type that has all values and all methods. Note that Any and builtin type object are completely different.

I don't interpret either of these as saying that `type(Any) is Any`. Also, mypy reports the type of `Any` as `object` for newer Python versions and `_AnyMeta` for older Python versions (as the definition of `Any` changed).

**Comment by erictraut** [comment]
You're confusing static type checking concepts with the runtime implementation of the special forms used for type checking. For example, the type of the expression `list[int]` from the perspective of a static type checker is `type[list[int]]`, but from the perspective of the runtime is `GenericAlias`. Static types and runtime types are not the same. From a static type perspective, the type of `Any` is `Any`, and the type of `type[Any]` is also `Any`. The runtime details may differ.

If you're interested in understanding the static type system in more detail, the official specification can be found [here](https://typing.readthedocs.io/en/latest/spec/index.html).

**Comment by hexane360** [comment]
I'm *not* confusing runtime and static types. I'm just pointing out that mypy returns a different result in this instance. From a static perspective, I'd expect the type of Any to be `t.Type[Any]`, not Any/the bottom type.

The documentation you linked agrees with the sources I've quoted. I don't see anything that specified the type of Any, maybe I'm missing something?

Do you think mypy's behavior is against spec here? If so, I can open an issue in their repo.

Thanks for taking your time to answer my questions

**Comment by erictraut** [comment]
Ah, the question you're asking is different from the OP's. That threw me off.

It sounds like you're trying to use `Any` in a value expression?

`Any` is intended to be used in type expressions, and the typing spec defines its meaning only when it is used in that context. The expression `type(Any)` is not a valid type expression. It is a valid value expression, but it's not clear why anyone would ever use that value expression in practice. That's why I'm curious what you're trying to do here. It seems just as strange as calling `type` on other special forms like `type(Protocol)` or `type(Generic)`.

**Comment by hexane360** [comment]
The short answer is I'm writing a well typed data conversion library (similar to mashumaro or pydantic). This includes a function `convert(val, type)`, where type is a supported type for conversion. Ideally, we should be able to type hint `type` that it allows `t.Any`, but not any value.

This was easier with old versions of the python stdlib and pyright, where `t.Type[t.Any]` was AnyMeta. It's harder with newer python stdlib, where it's `object`. But it's downright impossible if t.Any has type `t.Any`.

In a perfect world, t.Any would be treated as a singleton, i.e. the sole member of a unique (unnamed) type. Probably this would be too much work however.

Thanks for the clarification

**Comment by erictraut** [comment]
You might be interested in [this discussion](https://discuss.python.org/t/typeform-spelling-for-a-type-annotation-object-at-runtime/51435/13).

**Comment by erictraut** [comment]
I've decided to change the evaluation of the single-argument form of the `type` constructor to produce `type[Unknown]` if `x` is `typing.Any` or a value whose type is `Any`. I think this makes more sense than the current behavior where it evaluates to `Any`. This change will be included in the next release.

**Comment by hexane360** [comment]
Thanks, that will be easier to work with!
The discussion is helpful as well. Thanks for the quick response.
